### PR TITLE
Allow to read openstack cloud provider config from a secret

### PR DESCRIPTION
**Pull request by Fedosin**
This patch introduces new changes to allow reading openstack cloud provider config from a secret.

This change includes:

- Ability to read metadata values for kubelet. Since the service does not have access to the secret to read the configuration, but it needs data to download (e.g. hostname or flavor), we are trying to get it from the metadata server.

- Deprecation of kubeConfig parameter. Now we read the file that was provided with --kubeconfig option.

/kind feature

```release-note
Allow to read openstack cloud provider config from a secret
```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go [hunk 1]
```diff
@@ -261,59 +268,63 @@ func configFromEnv() (cfg Config, ok bool) {
...
-	if err != nil {
-		return fmt.Errorf("failed to get kubernetes client: %v", err)
+func (os *OpenStack) setConfigFromSecret() error {
+	if os.secretLister == nil {
+		return fmt.Errorf("secret lister is not initialized")
 	}
 
-	secret, err := k8sClient.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	secretName := os.secretName
+	secretNamespace := os.secretNamespace
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:283]: why copy values? just access directly

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go [hunk 2]
```diff
@@ -261,59 +268,63 @@ func configFromEnv() (cfg Config, ok bool) {
...
+	secret, err := os.secretLister.Secrets(secretNamespace).Get(secretName)
 	if err != nil {
-		klog.Warningf("Cannot get secret %s in namespace %s. error: %q", secretName, secretNamespace, err)
+		klog.Errorf("Cannot get secret %s in namespace %s. error: %q", secretName, secretNamespace, err)
 		return err
 	}
 
+	cfg := &Config{}
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:290]: why have this here outside of the if clause if it is only used inside?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go [hunk 3]
```diff
@@ -680,6 +699,11 @@ func getAttachedInterfacesByID(client *gophercloud.ServiceClient, serviceID stri
 
 // Clusters is a no-op
 func (os *OpenStack) Clusters() (cloudprovider.Clusters, bool) {
+	err := os.ensureCloudProviderWasInitialized()
+	if err != nil {
+		return nil, false
+	}
+
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:706]: why does it matter to initialize here?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go [hunk 4]
```diff
@@ -44,10 +44,36 @@ const (
 	instanceShutoff = "SHUTOFF"
 )
 
+func (os *OpenStack) getCompute() *gophercloud.ServiceClient {
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:47]: where is this used?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go [hunk 5]
```diff
@@ -166,6 +178,60 @@ func getMetadataFromMetadataService(metadataVersion string) (*Metadata, error) {
...
+func getIntanceType() (string, error) {
+	klog.V(4).Infof("Attempting to fetch instance type from %s", instanceTypeURL)
+	resp, err := http.Get(instanceTypeURL)
+	if err != nil {
+		return "", fmt.Errorf("error fetching %s: %v", instanceTypeURL, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("unexpected status code when reading instance type from %s: %s", instanceTypeURL, resp.Status)
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go:190]: why assign to err and just return?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go [hunk 6]
```diff
@@ -166,6 +178,60 @@ func getMetadataFromMetadataService(metadataVersion string) (*Metadata, error) {
...
+func getNodeAddress(url string) (string, error) {
+	klog.V(4).Infof("Attempting to fetch instance address from %s", url)
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", fmt.Errorf("error fetching %s: %v", url, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("unexpected status code when reading instance address from %s: %s", url, resp.Status)
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go:210]: why assign to err and just return?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go [hunk 7]
```diff
@@ -261,59 +268,63 @@ func configFromEnv() (cfg Config, ok bool) {
...
 	if content, ok := secret.Data["clouds.conf"]; ok {
 		err = gcfg.ReadStringInto(cfg, string(content))
 		if err != nil {
 			klog.Error("Cannot parse data from the secret.")
 			return fmt.Errorf("cannot parse data from the secret")
 		}
+		provider, err := newProvider(*cfg)
+		if err != nil {
+			klog.Errorf("Cannot initialize cloud provider using data from the secret.")
+			return fmt.Errorf("cannot initialize cloud provider using data from the secret")
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:301]: this has smell: return and logging an error

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go [hunk 8]
```diff
@@ -428,22 +431,37 @@ func newOpenStack(cfg Config) (*OpenStack, error) {
...
+		provider, err := newProvider(cfg)
+		if err != nil {
+			return nil, err
+		}
+		os.provider = provider
 	}
 
-	err = checkOpenStackOpts(&os)
+	err := checkOpenStackOpts(&os)
 	if err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:465]: `if err := ...; err != nil {`

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go [hunk 9]
```diff
@@ -44,10 +44,36 @@ const (
...
+
 // Instances returns an implementation of Instances for OpenStack.
 func (os *OpenStack) Instances() (cloudprovider.Instances, bool) {
 	klog.V(4).Info("openstack.Instances() called")
 
+	err := os.ensureCloudProviderWasInitialized()
+	if err != nil {
+		// cannot initialize cloud provider - return empty instances without compute instance,
+		// it will be generated later with any call of an Instances' receiver.
+		klog.Errorf("cannot initialize cloud provider, only limited functionality is available : %v", err)
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:71]: how often is this func called?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go [hunk 10]
```diff
@@ -166,6 +230,17 @@ func (os *OpenStack) InstanceID() (string, error) {
...
 // InstanceID returns the cloud provider ID of the specified instance.
 func (i *Instances) InstanceID(ctx context.Context, name types.NodeName) (string, error) {
+	// check if the node is local, in this case we can get its ID from the metadata service
+	// without additional requests to Nova.
+	md, err := getMetadata(i.opts.SearchOrder)
+	if err != nil {
+		return "", err
+	}
+	localName := types.NodeName(md.Name)
+	if localName == name {
```
Comment (sttts) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:240]: `if localName := types.NodeName(md.Name); localName == name {`

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go (hunk 1 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:283]: Done

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go (hunk 2 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:290]: Done

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go (hunk 3 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:706]: yep, I did it for consistency, but there is no need to initialize here.

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go (hunk 4 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:47]: nowhere, removed

File: staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go (hunk 5 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go:190]: Done

File: staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go (hunk 6 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/metadata.go:210]: Done

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go (hunk 7 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:301]: Done

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go (hunk 8 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:465]: Done

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go (hunk 9 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:71]: Instances? I don't know but quite often I think. Why do you ask?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go (hunk 10 above)
Comment (Fedosin) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:240]: Done

**Comment by marun** [comment]
@Fedosin Any update?

**Comment by soltysh** [comment]
@Fedosin any chance for update on this?

**Pull request closed without merging**